Core/Character.cs

```diff
@@ -1,105 +1,116 @@
+
+        public CharacterLevel WithVariable(string name, CharacterVariable value)
+        {
+            return new CharacterLevel(name, Rules, Variables.SetItem(name, value));
+        }
+
+        public BaseGameRules Rules { get; }
+
+        public CharacterVariable GetVariable(string name)
+        {
+            if (Variables.TryGetValue(name, out var result))
+                return result;
+            return new CharacterVariable(Rules);
+        }
+
+        public BonusBuilder<CharacterLevel> ModifyVariable(string name)
+        {
+            // This doesn't need to do anything, because "GetReference" will always create a variable that doesn't exist
+            // and set it to zero.  It's just a convenience to callers to get a BonusBuilder
+            return BonusBuilder.Create(this, name);
+        }
+    }
+
+    public class Character : IImmutableVariableCollection<Character>
     {
         public Character(BaseGameRules rules)
         {
             Rules = rules;
         }
 
+        private Character(BaseGameRules rules, string name, ImmutableList<CharacterLevel> levels, ImmutableDictionary<string, CharacterVariable> variables)
+        {
+            Rules = rules;
+            Name = name;
+            Levels = levels;
+            Variables = variables;
+        }
+
         public BaseGameRules Rules { get; }
 
-        public string? Name { get; set; }
+        public string Name { get; } = "";
+        
+        public Character WithName(string name)
+        {
+            return new Character(Rules, name, Levels, Variables);
+        }
+
+        public Character WithLevels(ImmutableList<CharacterLevel> levels)
+        {
+            return new Character(Rules, Name, levels, Variables);
+        }
+
+        public Character AddLevel(CharacterLevel level)
+        {
+            return new Character(Rules, Name, Levels.Add(level), Variables);
+        }
+
+        public Character WithVariable(string name, CharacterVariable value)
+        {
+            return new Character(Rules, Name, Levels, Variables.SetItem(name, value));
+        }
+
+        private ImmutableDictionary<string, CharacterVariable> Variables { get; } =
+            ImmutableDictionary.Create<string, CharacterVariable>(StringComparer.InvariantCultureIgnoreCase);
 
-        public Dictionary<string, CharacterVariable> Variables { get; } = new Dictionary<string, CharacterVariable>(StringComparer.InvariantCultureIgnoreCase);
+        public ImmutableList<CharacterLevel> Levels { get; } = ImmutableList<CharacterLevel>.Empty;
 
         public CharacterVariable GetVariable(string name)
         {
             if (Variables.TryGetValue(name, out var result))
                 return result;
-            result = new CharacterVariable(this, new List<Bonus>());
+            result = new CharacterVariable(Rules);
             Variables.Add(name, result);
```

Comment:
CharacterLevel isn't doing this add bit. Does it need to do this also?

Reply:
No, because it can't, because the Variables is an immutable dictionary.  This used to cache the "zero" so that everyone had the same variable reference (in case it got modified), now that they are all immutable, that's not an option.